File: scripts/backend-node/razorpay-migration/webhookRouter.js
/**
 * One webhook endpoint, two verifiers.
 *
 * Razorpay POSTs with the signature in `X-Razorpay-Signature` header; Paytm
 * POSTs with the signature in `body.head.signature`. We detect by shape.
 *
 * Mount with raw body parser BEFORE any JSON middleware - both verifiers
 * require the exact byte sequence Paytm/Razorpay signed over.
 */

import crypto from "node:crypto";
import PaytmChecksum from "paytmchecksum";

const RAZORPAY_WEBHOOK_SECRET = process.env.RAZORPAY_WEBHOOK_SECRET;
const PAYTM_MERCHANT_KEY = process.env.PAYTM_MERCHANT_KEY;

// Simple in-memory dedupe. Replace with Redis SETEX in production.
const recent = new Map();
const SEEN_TTL_MS = 10 * 60 * 1000;
function dedup(key) {
  const now = Date.now();
  for (const [k, t] of recent) if (now - t > SEEN_TTL_MS) recent.delete(k);
  if (recent.has(key)) return true;
  recent.set(key, now);
  return false;
}

function verifyRazorpay(rawBody, signature) {
  if (!signature || !RAZORPAY_WEBHOOK_SECRET) return false;
  const expected = crypto
    .createHmac("sha256", RAZORPAY_WEBHOOK_SECRET)
    .update(rawBody)
    .digest("hex");
  // timingSafeEqual would be ideal here in real code.
  return expected === signature;
}

async function verifyPaytm(rawBody, signature) {
  if (!signature) return false;
  return await PaytmChecksum.verifySignature(rawBody, PAYTM_MERCHANT_KEY, signature);
}

/** Express handler. Pass `express.raw({ type: "*/*" })` as middleware. */
export async function handleWebhook(req, res) {
  const rawBody = req.body.toString("utf8");
  let parsed;
  try {
    parsed = JSON.parse(rawBody);
  } catch {
    return res.status(400).send("invalid json");
  }

  const razorpaySig = req.get("X-Razorpay-Signature");
  const paytmSig = parsed?.head?.signature;

  let psp, ok, event;
  if (razorpaySig) {
    psp = "razorpay";
    ok = verifyRazorpay(rawBody, razorpaySig);
    event = parsed.event;
  } else if (paytmSig) {
    psp = "paytm";
    ok = await verifyPaytm(rawBody, paytmSig);
    event = `${parsed?.body?.txnType || "SALE"}.${parsed?.body?.status || "UNKNOWN"}`;
  } else {
    return res.status(401).send("no signature");
  }
  if (!ok) return res.status(401).send("invalid signature");

  // Dedup
  const dedupKey =
    psp === "razorpay"
      ? `razorpay:${req.get("X-Razorpay-Event-Id") || crypto.createHash("sha1").update(rawBody).digest("hex")}`
      : parsed?.body?.txnType === "REFUND"
        ? `paytm:refund:${parsed.body.refId}:${parsed.body.status}`
        : `paytm:${parsed.body.orderId}:${parsed.body.status}`;
  if (dedup(dedupKey)) return res.status(200).send("duplicate");

  try {
    await fulfillEvent({ psp, event, parsed, raw: rawBody });
    return res.status(200).send("ok");
  } catch (err) {
    console.error("[webhook]", psp, "fulfilment error", err);
    return res.status(500).send("retry");
  }
}

/**
 * Normalize and dispatch. Replace with your real business logic
 * (DB write, queue push, email, etc.).
 */
async function fulfillEvent({ psp, event, parsed }) {
  if (psp === "razorpay") {
    // Razorpay payload shape: { event, payload: { payment: { entity: { id, order_id, amount, status, ... } } } }
    const payment = parsed?.payload?.payment?.entity;
    if (!payment) return;
    // ... await db.updateOrderFromRazorpay(payment)
    console.log("[fulfil][razorpay]", event, payment.order_id, payment.status);
  } else {
    // Paytm payload shape: { head, body: { orderId, txnId, status, txnType, refId?, refundId?, ... } }
    const body = parsed?.body || {};
    if (body.txnType === "REFUND") {
      // ... await db.updateRefund(body)
      console.log("[fulfil][paytm][refund]", body.refId, body.status);
    } else {
      // ... await db.updateOrderFromPaytm(body)
      console.log("[fulfil][paytm]", body.orderId, body.status);
    }
  }
}

File: scripts/backend-python/razorpay_migration/webhook_router.py
"""
Razorpay + Paytm webhook router. One endpoint, two verifiers.

Razorpay -> signature in `X-Razorpay-Signature` header.
Paytm    -> signature in `body.head.signature` field.

Detect by shape and dispatch.
"""

from __future__ import annotations

import hashlib
import hmac
import json
import os
import time

from flask import Blueprint, request

from paytmchecksum import PaytmChecksum

webhook_bp = Blueprint("webhook", __name__)

# in-memory dedupe; swap for redis in prod
_seen: dict[str, float] = {}
_TTL = 600.0  # 10 min


def _dedup(key: str) -> bool:
    now = time.time()
    expired = [k for k, t in _seen.items() if now - t > _TTL]
    for k in expired:
        _seen.pop(k, None)
    if key in _seen:
        return True
    _seen[key] = now
    return False


def _verify_razorpay(raw_body: bytes, signature: str) -> bool:
    secret = os.getenv("RAZORPAY_WEBHOOK_SECRET")
    if not signature or not secret:
        return False
    expected = hmac.new(secret.encode("utf-8"), raw_body, hashlib.sha256).hexdigest()
    return hmac.compare_digest(expected, signature)


def _verify_paytm(raw_body_str: str, signature: str) -> bool:
    if not signature:
        return False
    return PaytmChecksum.verifySignature(raw_body_str, os.getenv("PAYTM_MERCHANT_KEY"), signature)


def _fulfill(psp: str, event: str, parsed: dict):
    """Replace with your business logic."""
    if psp == "razorpay":
        payment = (parsed.get("payload") or {}).get("payment", {}).get("entity") or {}
        print(f"[fulfil][razorpay] {event} {payment.get('order_id')} {payment.get('status')}")
    else:
        body = parsed.get("body") or {}
        if body.get("txnType") == "REFUND":
            print(f"[fulfil][paytm][refund] {body.get('refId')} {body.get('status')}")
        else:
            print(f"[fulfil][paytm] {body.get('orderId')} {body.get('status')}")


@webhook_bp.post("/paytm/webhook")
def handle_webhook():
    raw_body = request.get_data()
    raw_body_str = raw_body.decode("utf-8")
    try:
        parsed = json.loads(raw_body_str)
    except json.JSONDecodeError:
        return "invalid json", 400

    razorpay_sig = request.headers.get("X-Razorpay-Signature")
    paytm_sig = (parsed.get("head") or {}).get("signature")

    if razorpay_sig:
        psp = "razorpay"
        if not _verify_razorpay(raw_body, razorpay_sig):
            return "invalid signature", 401
        event = parsed.get("event", "unknown")
        dedup_key = (
            f"razorpay:{request.headers.get('X-Razorpay-Event-Id') or hashlib.sha1(raw_body).hexdigest()}"
        )
    elif paytm_sig:
        psp = "paytm"
        if not _verify_paytm(raw_body_str, paytm_sig):
            return "invalid signature", 401
        body = parsed.get("body") or {}
        event = f"{body.get('txnType') or 'SALE'}.{body.get('status') or 'UNKNOWN'}"
        if body.get("txnType") == "REFUND":
            dedup_key = f"paytm:refund:{body.get('refId')}:{body.get('status')}"
        else:
            dedup_key = f"paytm:{body.get('orderId')}:{body.get('status')}"
    else:
        return "no signature", 401

    if _dedup(dedup_key):
        return "duplicate", 200

    try:
        _fulfill(psp, event, parsed)
        return "ok", 200
    except Exception as e:  # noqa: BLE001
        print(f"[webhook] {psp} fulfilment error: {e}")
        return "retry", 500
